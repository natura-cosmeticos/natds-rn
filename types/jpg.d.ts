declare module '*.jpg' {
  const content: string
  export default content
}
