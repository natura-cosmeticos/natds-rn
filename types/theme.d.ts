import { Theme } from '../src/common/themeSelectors/theme/theme';

declare module 'styled-components' {
  interface DefaultTheme extends Theme {}
}
