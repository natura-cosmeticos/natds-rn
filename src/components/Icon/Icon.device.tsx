import { storiesOf } from '@storybook/react-native';
import { all, interactive } from './Icon.stories';

storiesOf('Icon', module)
  .add('all', all)
  .add('interactive', interactive);
