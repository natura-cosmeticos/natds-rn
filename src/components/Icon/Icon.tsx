import React from 'react';
import { withTheme } from 'styled-components/native';
import { Text } from 'react-native';
import iconNames from '@naturacosmeticos/natds-icons/dist/natds-icons.json';
import { ISizes } from '@naturacosmeticos/natds-styles';
import { Theme, IColors, getColorByName } from '../../common/themeSelectors';

export type IconColors = keyof IColors | 'default' | '#333333'
export type IconSizes = keyof ISizes

export interface IconProps {
  /**
   * An accessibility hint helps users understand what will happen when they perform an action
   * on the accessibility element when that result is not clear from the accessibility label.
   */
  accessibilityHint?: string,
  /**
   * Overrides the text that's read by the screen reader when the user interacts with the element.
   * By default, the label is constructed by traversing all the children and accumulating
   * all the Text nodes separated by space.
   */
  accessibilityLabel?: string
  /**
   * Icon color tokens
   */
  color?: IconColors,
  /**
   * Icon name
   */
  name?: string
  /**
   * Icon size
   */
  size?: IconSizes
  /**
   * Optional ID for testing
   */
  testID?: string,
  /**
   * The theme
   */
  theme: Theme,
}

const defaultIconName = 'outlined-default-mockup';

const getIconSize = (theme: Theme, size: IconSizes) => theme.sizes[size];

const getFontColor = (theme: Theme, color: IconColors) => {
  const colorName = color === 'default' ? 'highEmphasis' : color;

  return getColorByName(theme, colorName);
};

const IconComponent = ({
  accessibilityHint,
  accessibilityLabel,
  color = 'default',
  name = defaultIconName,
  testID = `icon-${name}`,
  theme,
  size = 'standard',
}: IconProps) => {
  const unicodeName = iconNames[name]
    ? iconNames[name].replace('%', '\\')
    : iconNames[defaultIconName];

  const code = JSON.parse(`["${unicodeName}"]`)[0];

  return (
    <Text
      accessibilityHint={accessibilityHint}
      accessibilityLabel={accessibilityLabel}
      accessibilityRole="text"
      style={{
        color: color === '#333333' ? color : getFontColor(theme, color),
        fontFamily: 'natds-icons',
        fontSize: getIconSize(theme, size),
      }}
      testID={testID}
    >
      {code}
    </Text>
  );
};

export const Icon = withTheme(IconComponent);
