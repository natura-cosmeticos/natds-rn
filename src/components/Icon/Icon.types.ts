import { IconName } from '@naturacosmeticos/natds-icons'
import { Color, Size } from '@naturacosmeticos/natds-themes/react-native'
import { StyleProp, TextStyle } from 'react-native'
import { Theme } from '../..'

/**
 * @deprecated Use only natds-themes color tokens
 */
type IconsColorsDeprecated = 'default' | '#333333'
export type IconColors = keyof Color | IconsColorsDeprecated
export type IconSizes = keyof Size

export interface IconProps {
  /**
   * An accessibility hint helps users understand what will happen when they perform an action
   * on the accessibility element when that result is not clear from the accessibility label.
   */
  accessibilityHint?: string;
  /**
   * Overrides the text that's read by the screen reader when the user interacts with the element.
   * By default, the label is constructed by traversing all the children and accumulating
   * all the Text nodes separated by space.
   */
  accessibilityLabel?: string;
  /**
   * Communicates the purpose of a component to the user of an assistive technology
   */
  accessibilityRole?: 'imagebutton' | 'image';
  /**
   * Icon color tokens
   * #### deprecated colors: `default` and `#333333`:
   * Use natds-themes color tokens instead
   */
  color?: IconColors;
  /**
   * Icon name
   */
  name?: IconName;
  /**
   * Icon size
   */
  size?: IconSizes;
  /**
   * Optional ID for testing
   */
  testID?: string;
  /**
   * The theme
   */
  theme?: Theme;
  /**
   *
   */
  style?: StyleProp<TextStyle>;
}
