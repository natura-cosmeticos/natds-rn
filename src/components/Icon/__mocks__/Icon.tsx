import React from 'react'
import { Text } from 'react-native'

export const Icon = (props) => {
  const { name } = props
  return (
    <Text {...props}>
      {name}
    </Text>
  )
}
