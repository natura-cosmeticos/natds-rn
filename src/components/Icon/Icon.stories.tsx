import React from 'react';
import { text as textKnob } from '@storybook/addon-knobs';
import { IconName } from '@naturacosmeticos/natds-icons';
import { Icon } from './Icon';
import { IconColors, IconSizes } from './Icon.types';
import { StoryContainer } from '../../common/HelperComponents/StoryContainer';

const description = () => `
---

**NOTE**: This component is available with the following attribute status:

- ✅ **Color:** all supported Color tokens from \`natds-themes\`
- ✅  **Size:** all supported Size tokens from \`natds-themes\`
- ✅  **Name:** all supported Icon names from [Icon Library](https://ds.natura.design/28db352be/p/94367e-icon-library/b/6154b9)

---
`;

export default {
  component: Icon,
  parameters: {
    componentSubtitle: 'A helper component to display icons from @naturacosmeticos/natds-icons package',
    docs: {
      extractComponentDescription: description,
    },
  },
  title: 'Components/Icon',
};

export const Default = () => (
  <StoryContainer title="Default">
    <Icon />
  </StoryContainer>
);

export const Interactive = () => (
  <StoryContainer title="Interactive">
    <Icon
      color={textKnob('Color', 'highlight')as IconColors}
      name={textKnob('Icon name', 'outlined-default-mockup') as IconName}
      size={textKnob('Size', 'standard') as IconSizes}
    />
  </StoryContainer>
);
