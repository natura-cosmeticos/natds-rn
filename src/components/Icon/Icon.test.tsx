import { IconName } from '@naturacosmeticos/natds-icons';
import React from 'react';
import { checkIconName, Icon } from './Icon';
import { renderWithTheme } from '../../../test/testHelpers';
import { IconProps } from './Icon.types';

jest.mock('../../common/themeSelectors', () => (
  {
    getColorByName: () => '#BBBBBB',
    getSize: () => 24,
  }));

const renderIcon = (props?: Partial<IconProps>) => (
  renderWithTheme(<Icon {...props} />)
);

const sampleNames: Array<IconName> = [
  'filled-action-add',
  'outlined-action-micrec',
  'outlined-content-wellness',
  'outlined-finance-bank',
];

describe('Icon component', () => {
  it('should render component with default props', () => {
    const { toJSON } = renderIcon();

    expect(toJSON()).toMatchSnapshot('Icon component - name: outlined-default-mockup');
  });

  it('should render component with given props', () => {
    const { queryByTestId, toJSON } = renderIcon({
      accessibilityRole: 'imagebutton',
      name: 'outlined-finance-bank',
    });

    expect(toJSON()).toMatchSnapshot('Icon component - name: outlined-finance-bank');
    expect(queryByTestId('natds-icon')?.props).toHaveProperty('accessibilityRole', 'imagebutton');
  });

  it('should render deafult icon if the icon name is incorrect', () => {
    const defaultIconName = 'outlined-default-mockup';

    const anyIncorrectIconName = checkIconName('incorrect-icon-name' as IconName);
    const expectedIconName = checkIconName(defaultIconName);

    expect(anyIncorrectIconName).toBe(expectedIconName);
  });

  /* eslint-disable-next-line mocha/no-setup-in-describe */
  it.each(sampleNames)('should render correctly when icon is %p', (name) => {
    const { toJSON } = renderIcon({ name });

    expect(toJSON()).toMatchSnapshot(`Icon component - name: ${name}`);
  });
});
