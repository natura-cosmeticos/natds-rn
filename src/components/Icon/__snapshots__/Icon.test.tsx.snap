// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon component should render component with default props: Icon component - name: outlined-default-mockup 1`] = `
<Text
  accessibilityRole="image"
  style={
    Object {
      "color": "#BBBBBB",
      "fontFamily": "natds-icons",
      "fontSize": 24,
    }
  }
  testID="natds-icon"
>
  
</Text>
`;

exports[`Icon component should render component with given props: Icon component - name: outlined-finance-bank 1`] = `
<Text
  accessibilityRole="imagebutton"
  style={
    Object {
      "color": "#BBBBBB",
      "fontFamily": "natds-icons",
      "fontSize": 24,
    }
  }
  testID="natds-icon"
>
  
</Text>
`;

exports[`Icon component should render correctly when icon is "filled-action-add": Icon component - name: filled-action-add 1`] = `
<Text
  accessibilityRole="image"
  style={
    Object {
      "color": "#BBBBBB",
      "fontFamily": "natds-icons",
      "fontSize": 24,
    }
  }
  testID="natds-icon"
>
  
</Text>
`;

exports[`Icon component should render correctly when icon is "outlined-action-micrec": Icon component - name: outlined-action-micrec 1`] = `
<Text
  accessibilityRole="image"
  style={
    Object {
      "color": "#BBBBBB",
      "fontFamily": "natds-icons",
      "fontSize": 24,
    }
  }
  testID="natds-icon"
>
  
</Text>
`;

exports[`Icon component should render correctly when icon is "outlined-content-wellness": Icon component - name: outlined-content-wellness 1`] = `
<Text
  accessibilityRole="image"
  style={
    Object {
      "color": "#BBBBBB",
      "fontFamily": "natds-icons",
      "fontSize": 24,
    }
  }
  testID="natds-icon"
>
  
</Text>
`;

exports[`Icon component should render correctly when icon is "outlined-finance-bank": Icon component - name: outlined-finance-bank 1`] = `
<Text
  accessibilityRole="image"
  style={
    Object {
      "color": "#BBBBBB",
      "fontFamily": "natds-icons",
      "fontSize": 24,
    }
  }
  testID="natds-icon"
>
  
</Text>
`;
